[Tan Q] iP

tanqiya opened the pull request:
### Duck the Duke 🦆 

> "Hello! I'm Duck! *quack*" - Duck the Duke

When life gets messy, Duck's got your back! 
![Duck PR](https://user-images.githubusercontent.com/35829668/131251247-7a1259ed-bc96-41d0-a51a-722c9d76cd1c.jpg)

Duck is a chatbot that helps you manage your task list. It
- is easy to learn
- uses simple commands
- is very cute!!

Features:

- [ ] Adding tasks (`Todo`, `Deadline` and `Event`)
- [ ] Marking your tasks as completed
- [ ] Deleting tasks from your list

An example of using the `list` command to display your task list:

1. Read Week 4 Content
2. Complete Week 4 Quiz (by: Aug 30 2021)
3. Attend CS2103T Tutorial (at: Sep 1 2021)

View the code [here](https://github.com/tanqiya/ip).

Here's the deconstructed greeting in code from the image above:
```java
private static final String GREETING = "Hello! I'm Duck \n"
            + "*quack*  >(.)__\n"
            + "          (___/ \n"
            + "What can I do for you?\n";
```

Commit:
Add Gradle support

tanqiya committed:
Create Duck that echos user input. Level-1

tanqiya committed:
Add list to store user input

tanqiya committed:
Add completion status for task

tanqiya committed:
Add three types of tasks. Create Task class

tanqiya committed:
Create class for each type of task

tanqiya committed:
Add Exceptions for handling errors from using input

tanqiya committed:
Add delete function. Change task to ArrayList.

tanqiya committed:
Use I/O technique to semi-automate testing

tanqiya committed:
Save tasks to hard disk

tanqiya committed:
Store date as LocalDate

tanqiya committed:
Move task-related classes to task package

tanqiya committed:
Merge branch 'wrong-branch'

tanqiya committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/tasks/Deadline.java
#	src/main/java/tasks/Event.java

tanqiya committed:
Organize into packages and extract classes for OOP.

tanqiya committed:
Add JUnit tests for Deadline and Todo

tanqiya committed:
Create a JAR file

tanqiya committed:
Added JavaDoc header comments for main classes and public methods in duke package.

tanqiya committed:
Tweak code to comply with coding standard

tanqiya committed:
Add preliminary find function that matches task description

tanqiya committed:
Minor change to error formatting

tanqiya committed:
Merge branch 'branch-A-JavaDoc'

tanqiya committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java

tanqiya committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java

sunjc826 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    /**
+     * Opens the file on hard disk where user's past tasks were saved.
```
```suggestion
Opens the file on disk where user's past tasks were saved.
```
tiny nitpick

sunjc826 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,126 @@
+            Files.createDirectories(directoryPath); //Create directory if it does not exist
+            File taskRecord = new File(filepath);
+            if (!taskRecord.createNewFile()) {
+                //File already exists, parse file and add duke.tasks to ArrayList
```
```suggestion
// File already exists, parse file and add duke.tasks to ArrayList
```
I think inline comments look better with a space first.

sunjc826 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    /**
+     * Rewrite the entire file whenever a task is deleted or marked as done.
```
```suggestion
* Rewrite the entire file whenever a task is deleted or marked as done.

```

sunjc826 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,151 @@
+    public void showFindResults(ArrayList<Task> results) {
+        if (results.isEmpty()) {
+            System.out.println("Sorry, I did not find any task that matches your search *sad quack*\n");
+        } else {
+            System.out.println("Here are the tasks I found:");
+            for (Task task : results) {
+                System.out.println(task.toString());
+            }
+            System.out.println("*quack*\n");
+        }
```
```suggestion
            
            System.out.println("Here are the tasks I found:");
            for (Task task : results) {
                System.out.println(task.toString());
            }
            System.out.println("*quack*\n");
```
I think code can look nicer and clearer without the else.

sunjc826 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,151 @@
+            System.out.println("*quack*\n");
+        }
+    }
+}
```
Do remember to add a new line here at the end of each file.

sunjc826 commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,42 @@
+import duke.tasks.Todo;
+
+public class AddCommand extends Command {
+    private Duke.TaskType taskType;
```
If the `TaskType` class is to be used outside, should it still be placed inside `Duke`?
Additonally, is there a reason to couple `TaskType` with `Duke`?

sunjc826 commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,44 @@
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public void taskDone() {
```
Suggest rename to `markTaskDone`, since this is a verb.

sunjc826 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+     * @throws TaskOutOfRangeException if taskNo is out of range.
+     */
+    public void deleteTask(int taskNo) throws TaskOutOfRangeException {
+        if (taskNo < 1 || taskNo > taskList.size()) {
```
This technically violates "Avoid complicated expressions", but a single OR condition should be fine in practice.

sunjc826 reviewed (commented):
Overall, the code is very clean. There isn't much in the way of deep nesting of code, and that's great. Method names are generally descriptive, however, I do think a bit more documentation would be good!

wei2912 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+        tasks = new TaskList(storage.load());
+    }
+
+    public void run() {
```
As this is a public function, it would benefit from having a JavaDoc comment. (This issue appears in other files too.)

wei2912 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,98 @@
+package duke;
+
+import duke.commands.*;
```
As per the coding standard, all imported classes should be listed explicitly. I think this issue appears in a couple of other files too. You might want to configure your IDE if it is automatically converting multiple imports into wildcard imports.

wei2912 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,126 @@
+     * @param taskRecord record of user's past tasks to be parsed.
+     * @return list of tasks to be used to construct TaskList.
+     */
+    public static ArrayList<Task> readSaved (File taskRecord) {
```
Just a tiny nitpick, there's an additional space between `readSaved` and `(`.
```suggestion
    public static ArrayList<Task> readSaved(File taskRecord) {
```

wei2912 commented on src/main/java/duke/exceptions/TaskOutOfRangeException.java:
```diff
@@ -0,0 +1,23 @@
+    public TaskOutOfRangeException(int maxIndex) {
+        super(maxIndex == 1 
+                ? "Oops! No such task exists! You only have one task.\n"
+                :"Oops! No such task exists\nPlease use a number from 1 to " 
```
Just another nitpick with spacing.
```suggestion
                : "Oops! No such task exists\nPlease use a number from 1 to " 
```

wei2912 commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,30 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
+    protected LocalDate by;
```
While I don't think this explicitly violates the coding standard, I think it would be clearer to use a noun for this variable such as `date`.

wei2912 reviewed (commented):
Great work so far! Generally your code looks clear and concise (although I haven't looked through much of it), although there were a few minor coding standard violations.

tanqiya committed:
Respond to PR review comments. Change IDE code style.

tanqiya committed:
Merge remote-tracking branch 'origin/add-gradle-support'

tanqiya committed:
Add checkstyle to detect coding style violations

tanqiya committed:
Add basic GUI from JavaFX tutorial

tanqiya committed:
Link GUI with Duke commands

tanqiya committed:
Improve Code Quality

Look through code to find areas for improvement and improve readability.

Remove unnecessary else-blocks.

Make the happy path prominent according to Code Quality topics covered.

tanqiya committed:
Add Assertions

Checks variables to prevent potential bugs in code.

Add assertion lines to check user input and correct parsing.

tanqiya committed:
Add Level 10 branch for ip tracking

tanqiya committed:
Merge branch 'branch-Level-10'

tanqiya committed:
Merge pull request #1 from tanqiya/branch-A-CodeQuality

Improve Code Quality

tanqiya committed:
Merge branch 'master' into branch-A-Assertions

tanqiya committed:
Merge pull request #2 from tanqiya/branch-A-Assertions

Add Assertions

tanqiya committed:
Add better search.

Find can search for task descriptions that contain a substring of the search term. Would help with typos.

tanqiya committed:
Improve GUI

Fixed text overflow problem when list was too long. Added timer to close JavaFX after 'bye' command.

tanqiya committed:
Finalise Features

Created help function for user to view supported commands. Added Javadocs for public methods and JavaFX window title and image.

tanqiya committed:
Add User Guide and Ui.png

Also final changes to GUI messages

tanqiya committed:
Set theme jekyll-theme-minimal

tanqiya committed:
Add dialog box colour and change exception messages
